projects/core/src/lib/classes/mask-model/mask-model.ts

```diff
@@ -0,0 +1,88 @@
+import {MaskExpression} from '../../types';
+import {validateValueWithMask} from './utils/validate-value-with-mask';
+import {removeFixedMaskCharacters} from './utils/remove-fixed-mask-characters';
+import {addFixedMaskCharacters} from './utils/add-fixed-mask-characters';
+
+export class MaskModel {
+    value = '';
+    caretIndex = 0;
+
+    constructor(
+        readonly initialValue: string,
+        readonly initialCaretIndex: number,
+        private readonly mask: MaskExpression,
+    ) {
+        const {maskedValue, maskedCaretPosition} = addFixedMaskCharacters(
+            initialValue,
+            this.mask,
+            initialCaretIndex,
+        );
+
+        this.value = maskedValue;
+        this.caretIndex = maskedCaretPosition;
+    }
+
+    addCharacters(
+        selectedRange: [from: number, to: number],
+        newCharacters: string,
+    ): void {
+        const {unmaskedValue, unmaskedSelection} = removeFixedMaskCharacters(
+            this.value,
+            this.mask,
+            selectedRange,
+        );
+        const [unmaskedFrom, unmaskedTo] = unmaskedSelection;
+        const newUnmaskedValue =
+            unmaskedValue.slice(0, unmaskedFrom) +
+            newCharacters +
+            unmaskedValue.slice(unmaskedTo);
+
+        const {maskedValue, maskedCaretPosition} = addFixedMaskCharacters(
+            newUnmaskedValue,
+            this.mask,
+            unmaskedTo,
+        );
+
+        if (!validateValueWithMask(maskedValue, this.mask)) {
+            throw new Error('Invalid mask value');
+        }
+
+        this.value = maskedValue;
+        this.caretIndex = maskedCaretPosition;
+    }
+
+    removeCharacters([from, to]: [from: number, to: number]): void {
+        if (from === 0 && to === 0) {
+            return;
+        }
+
+        const selectedRange: [number, number] = from === to ? [from - 1, to] : [from, to];
+        const {unmaskedValue, unmaskedSelection} = removeFixedMaskCharacters(
+            this.value,
+            this.mask,
+            selectedRange,
+        );
+        const [unmaskedFrom, unmaskedTo] = unmaskedSelection;
+        const isOnlyFixedCharsSelected = unmaskedFrom === unmaskedTo;
+        const isLastChar = unmaskedTo >= unmaskedValue.length;
+
+        if (isOnlyFixedCharsSelected && !isLastChar) {
+            this.caretIndex = selectedRange[0];
+            return;
+        }
+
+        const newUnmaskedValue =
```

Comment:
This big ternary is hard to read, let's move it to a separate constant:
```ts
const newFrom = isOnlyFixedCharsSelected && isLastChar ? unmaskedFrom - 1 : unmaskedFrom;
```
Hope it fits in one line :D maybe `isOnlyFixedCharsSelected` can drop `is` or made shorter some other way to keep such things in one line — it's easier to follow this way :)